### PR TITLE
Add support for variable playback speed for Desktop Access recordings

### LINKED ISSUE #14430: Desktop Access Playback Speed
Support 2x/4x/8x playback speed in the web player.

### PR DESCRIPTION
**Pull request by xacrimon**
This PR introduces a new player action event to be received over websocket which adjusts the playback speed of the event streamer. This allows for variable playback speed in the client. This doesn't break anything without a webapps PR as it's purely an additional action that regulates how fast the events are sent to the frontend.

Requires https://github.com/gravitational/webapps/pull/1072 to enable this functionality in the frontend player.

Fixes https://github.com/gravitational/teleport/issues/14430

### CODE REVIEW

**Comment by xacrimon** [comment]
https://user-images.githubusercontent.com/21025159/182433299-8e0ebdb0-81b2-4af2-82ae-181ca957945f.mov

File: lib/web/desktop/playback.go [hunk 1]
```diff
@@ -219,6 +230,10 @@ func (pp *Player) streamSessionEvents(ctx context.Context, cancel context.Cancel
 				}
 			}
 			return
+		case playSpeed := <-pp.playSpeed:
+			prevDelayFactor := delayFactor
+			delayFactor = 1.0 / playSpeed
+			lastDelay = int64(float32(lastDelay) * (delayFactor / prevDelayFactor))
```
Comment (zmb3) [inline lib/web/desktop/playback.go:236]: Let's keep `lastDelay` equivalent to the `Delay` field of the _last_ event we played.

Why can't we just divide by the playback speed where we do the sleep and leave `lastDelay` alone?

Comment (xacrimon) [reply lib/web/desktop/playback.go]: This did grow more complicated than it needed to be as I cycled though a couple approaches. I just simplified it quite a bit.

**Comment by zmb3** [comment]
The simplifications look good. Should we also restrict to a known set of speeds rather than allowing any value the API sends?

**Comment by xacrimon** [comment]
That seems reasonable but I don't like that we end up defining all possible speeds in a list at two places IMO. I added min and max settings to make things easy to adjust in the future but keeping one big list in two places seems like asking for trouble. The min/max limits should prevent any misuse.

**Review by ibeckermayer — CHANGES_REQUESTED** [review]
An issue arises if you pause the playback, change speeds, and then resume:

https://user-images.githubusercontent.com/13578537/182479894-a92cdaf8-e1ae-4e3c-8423-bbc8eac8201f.mov

The playback never starts playing back again.

What's happening is that when you pause the playback, `streamSessionEvents` hangs here
https://github.com/gravitational/teleport/blob/4a488732eaaeefda5f5cce66389dda1a425aa857/lib/web/desktop/playback.go#L226

and then when you change the playback speed, `receiveActions` starts hanging here
https://github.com/gravitational/teleport/blob/4a488732eaaeefda5f5cce66389dda1a425aa857/lib/web/desktop/playback.go#L208

Because that `streamSessionEvents` loop is hanging at `pp.waitWhilePaused()` you never make it down to 
https://github.com/gravitational/teleport/blob/4a488732eaaeefda5f5cce66389dda1a425aa857/lib/web/desktop/playback.go#L245
to "un-hang" the `receiveActions` loop on its `pp.playSpeed <- action.PlaybackSpeed` channel write. This means that when you try to "unpause" (send another `actionPlayPause`, meant to be handled by `receiveActions`), it never gets handled.

I think it would solve this if instead of using a channel, we made `playSpeed` a mutex-guarded `float32`.

**Comment by xacrimon** [comment]
@ibeckermayer thanks! I pushed a commit that uses `pp.mu` instead and can confirm pausing, changing the speed and resuming works now.

**Review by ibeckermayer — APPROVED** [review]

**Review by zmb3 — APPROVED** [review]

**Pull request merged**